// @gfazioli/mantine-window/dist/esm/Window.mjs
'use client';
import React from 'react';
import { IconX, IconPlus, IconMinus } from '@tabler/icons-react';
import { createVarsResolver, getShadow, getRadius, factory, useProps, useStyles, Box, Flex, ActionIcon, Text, ScrollArea } from '@mantine/core';
import { useMantineWindow } from './hooks/use-mantine-window.mjs';
import classes from './Window.module.css.mjs';

const defaultProps = {
  withBorder: true,
  shadow: "md",
  resizable: "both",
  draggable: "both",
  collapsed: false,
  withCloseButton: true,
  withCollapseButton: true,
  collapsable: true,
  persistState: false,
  withinPortal: true,
  minWidth: 250,
  minHeight: 100
};
const varsResolver = createVarsResolver((_, { radius, shadow }) => {
  return {
    root: {
      "--window-background": "var(--mantine-color-default)",
      "--window-radius": radius === void 0 ? "var(--mantine-radius-lg)" : getRadius(radius),
      "--window-shadow": getShadow(shadow)
    },
    container: {},
    content: {},
    header: {},
    title: {},
    closeButton: {},
    collapseButton: {},
    windowToolsButton: {},
    resizeHandleTopLeft: {},
    resizeHandleTop: {},
    resizeHandleTopRight: {},
    resizeHandleRight: {},
    resizeHandleBottomRight: {},
    resizeHandleBottom: {},
    resizeHandleBottomLeft: {},
    resizeHandleLeft: {}
  };
});
const Window = factory((_props, _) => {
  const props = useProps("Window", defaultProps, _props);
  const {
    opened,
    color,
    title,
    children,
    resizable,
    draggable,
    collapsed,
    withCollapseButton,
    collapsable,
    withCloseButton,
    onClose,
    radius,
    withinPortal,
    persistState,
    minWidth,
    minHeight,
    maxWidth,
    maxHeight,
    dragBounds,
    defaultPosition,
    defaultSize,
    onPositionChange,
    onSizeChange,
    withBorder,
    mod,
    classNames,
    style,
    styles,
    unstyled,
    vars,
    className,
    ...others
  } = props;
  const getStyles = useStyles({
    name: "Window",
    props,
    classes,
    className,
    style,
    classNames,
    styles,
    unstyled,
    vars,
    varsResolver
  });
  const {
    bringToFront,
    windowRef,
    position,
    size,
    isCollapsed,
    setIsCollapsed,
    isVisible,
    zIndex,
    handleMouseDownDrag,
    handleTouchStartDrag,
    resizeHandlers,
    handleClose
  } = useMantineWindow(props);
  const draggableHeader = draggable === "header" || draggable === "both";
  const draggableWindow = draggable === "window" || draggable === "both";
  if (!isVisible) {
    return null;
  }
  return /* @__PURE__ */ React.createElement(
    Box,
    {
      ref: windowRef,
      onClick: bringToFront,
      mod: [{ "data-with-border": withBorder, "data-window-draggable": draggableWindow }, mod],
      onMouseDown: draggableWindow ? handleMouseDownDrag : void 0,
      onTouchStart: draggableWindow ? handleTouchStartDrag : void 0,
      bg: color,
      ...others,
      ...getStyles("root", {
        style: {
          position: withinPortal ? "fixed" : "absolute",
          left: position.x,
          top: position.y,
          width: size.width,
          height: isCollapsed ? "auto" : size.height,
          zIndex
        }
      })
    },
    /* @__PURE__ */ React.createElement(Box, { ...getStyles("container") }, /* @__PURE__ */ React.createElement(
      Box,
      {
        ...getStyles("header"),
        onClick: bringToFront,
        mod: { "window-draggable": draggableHeader },
        onMouseDown: draggableHeader ? handleMouseDownDrag : void 0,
        onTouchStart: draggableHeader ? handleTouchStartDrag : void 0,
        onDoubleClick: () => collapsable && setIsCollapsed(!isCollapsed)
      },
      /* @__PURE__ */ React.createElement(Flex, { align: "center", gap: "xs", miw: 0 }, /* @__PURE__ */ React.createElement(Flex, { align: "center", gap: 8 }, withCloseButton && /* @__PURE__ */ React.createElement(
        ActionIcon,
        {
          radius: 256,
          color: "red",
          onClick: handleClose,
          ...getStyles("closeButton")
        },
        /* @__PURE__ */ React.createElement(IconX, { size: 14 })
      ), withCollapseButton && collapsable && /* @__PURE__ */ React.createElement(
        ActionIcon,
        {
          radius: 256,
          color: "yellow",
          onClick: () => setIsCollapsed(!isCollapsed),
          ...getStyles("collapseButton")
        },
        isCollapsed ? /* @__PURE__ */ React.createElement(IconPlus, { size: 14 }) : /* @__PURE__ */ React.createElement(IconMinus, { size: 14 })
      )), /* @__PURE__ */ React.createElement(Flex, { align: "center", gap: "xs", miw: 0 }, /* @__PURE__ */ React.createElement(Text, { truncate: true, ...getStyles("title") }, title)))
    ), !isCollapsed && /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
      ScrollArea,
      {
        ...getStyles("content", {
          style: {
            height: size.height - 40
          }
        })
      },
      children
    ), resizable !== "none" && /* @__PURE__ */ React.createElement(React.Fragment, null, resizable === "both" && /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
      Box,
      {
        "data-resize-handle": true,
        ...resizeHandlers.topLeft,
        ...getStyles("resizeHandleTopLeft")
      }
    ), /* @__PURE__ */ React.createElement(
      Box,
      {
        "data-resize-handle": true,
        ...resizeHandlers.topRight,
        ...getStyles("resizeHandleTopRight")
      }
    ), /* @__PURE__ */ React.createElement(
      Box,
      {
        "data-resize-handle": true,
        ...resizeHandlers.bottomRight,
        ...getStyles("resizeHandleBottomRight")
      }
    ), /* @__PURE__ */ React.createElement(
      Box,
      {
        "data-resize-handle": true,
        ...resizeHandlers.bottomLeft,
        ...getStyles("resizeHandleBottomLeft")
      }
    )), (resizable === "vertical" || resizable === "both") && /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
      Box,
      {
        "data-resize-handle": true,
        ...resizeHandlers.top,
        ...getStyles("resizeHandleTop")
      }
    ), /* @__PURE__ */ React.createElement(
      Box,
      {
        "data-resize-handle": true,
        ...resizeHandlers.bottom,
        ...getStyles("resizeHandleBottom")
      }
    )), (resizable === "horizontal" || resizable === "both") && /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
      Box,
      {
        "data-resize-handle": true,
        ...resizeHandlers.right,
        ...getStyles("resizeHandleRight")
      }
    ), /* @__PURE__ */ React.createElement(
      Box,
      {
        "data-resize-handle": true,
        ...resizeHandlers.left,
        ...getStyles("resizeHandleLeft")
      }
    )))))
  );
});
Window.classes = classes;
Window.displayName = "Window";

export { Window, defaultProps };
