// @gfazioli/mantine-window/dist/esm/hooks/use-mantine-window.mjs
'use client';
import { useState, useCallback, useRef, useEffect, useMemo } from 'react';
import { useLocalStorage } from '@mantine/hooks';

const CURSOR_MAP = {
  topLeft: "nwse-resize",
  top: "ns-resize",
  topRight: "nesw-resize",
  right: "ew-resize",
  bottomRight: "nwse-resize",
  bottom: "ns-resize",
  bottomLeft: "nesw-resize",
  left: "ew-resize"
};
function useMantineWindow(props) {
  const {
    title,
    collapsed,
    opened,
    onClose,
    id,
    persistState = false,
    withinPortal = true,
    defaultPosition = { x: 20, y: 100 },
    defaultSize = { width: 400, height: 400 },
    minWidth = 250,
    minHeight = 100,
    maxWidth,
    maxHeight,
    dragBounds,
    onPositionChange,
    onSizeChange
  } = props;
  const [isCollapsed, setIsCollapsed] = useState(collapsed ?? false);
  const [isVisible, setIsVisible] = useState(opened ?? false);
  const [zIndex, setZIndex] = useState(200);
  const key = (id || title)?.toLocaleLowerCase().replace(/\s+/g, "-") || "window";
  const [positionStorage, setPositionStorage] = useLocalStorage({
    key: persistState && `${key}-window-position`,
    defaultValue: defaultPosition,
    getInitialValueInEffect: true
  });
  const [sizeStorage, setSizeStorage] = useLocalStorage({
    key: persistState && `${key}-window-size`,
    defaultValue: defaultSize,
    getInitialValueInEffect: true
  });
  const [positionState, setPositionState] = useState(defaultPosition);
  const [sizeState, setSizeState] = useState(defaultSize);
  const position = persistState ? positionStorage : positionState;
  const size = persistState ? sizeStorage : sizeState;
  const setPosition = useCallback(
    (newPosition) => {
      const pos = typeof newPosition === "function" ? newPosition(position) : newPosition;
      if (persistState) {
        setPositionStorage(pos);
      } else {
        setPositionState(pos);
      }
      onPositionChange?.(pos);
    },
    [persistState, position, setPositionStorage, onPositionChange]
  );
  const setSize = useCallback(
    (newSize) => {
      const sz = typeof newSize === "function" ? newSize(size) : newSize;
      if (persistState) {
        setSizeStorage(sz);
      } else {
        setSizeState(sz);
      }
      onSizeChange?.(sz);
    },
    [persistState, size, setSizeStorage, onSizeChange]
  );
  const windowRef = useRef(null);
  const isDragging = useRef(false);
  const isResizing = useRef(false);
  const dragStart = useRef({ x: 0, y: 0 });
  const resizeStart = useRef({ x: 0, y: 0, width: 0, height: 0, posX: 0, posY: 0 });
  useEffect(() => {
    setIsVisible(opened ?? false);
  }, [opened]);
  useEffect(() => {
    setIsCollapsed(collapsed ?? false);
  }, [collapsed]);
  useEffect(() => {
    const handleClickOutside = (event) => {
      if (windowRef.current && !windowRef.current.contains(event.target)) {
        setZIndex(199);
      }
    };
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);
  const bringToFront = useCallback(() => {
    setZIndex(200);
  }, []);
  const handleMouseDownDrag = useCallback(
    (e) => {
      if (e.target.closest("[data-resize-handle]")) {
        return;
      }
      bringToFront();
      isDragging.current = true;
      dragStart.current = {
        x: e.clientX - position.x,
        y: e.clientY - position.y
      };
      document.body.style.userSelect = "none";
      e.preventDefault();
    },
    [position, bringToFront]
  );
  const handleTouchStartDrag = useCallback(
    (e) => {
      if (e.target.closest("[data-resize-handle]")) {
        return;
      }
      const touch = e.touches[0];
      bringToFront();
      isDragging.current = true;
      dragStart.current = {
        x: touch.clientX - position.x,
        y: touch.clientY - position.y
      };
      document.body.style.userSelect = "none";
      e.preventDefault();
    },
    [position, bringToFront]
  );
  const resizeDirection = useRef("");
  const createResizeHandlers = useCallback(
    (direction) => {
      const onMouseDown = (e) => {
        bringToFront();
        isResizing.current = true;
        resizeDirection.current = direction;
        resizeStart.current = {
          x: e.clientX,
          y: e.clientY,
          width: size.width,
          height: size.height,
          posX: position.x,
          posY: position.y
        };
        document.body.style.cursor = CURSOR_MAP[direction];
        document.body.style.userSelect = "none";
        e.preventDefault();
        e.stopPropagation();
      };
      const onTouchStart = (e) => {
        const touch = e.touches[0];
        bringToFront();
        isResizing.current = true;
        resizeDirection.current = direction;
        resizeStart.current = {
          x: touch.clientX,
          y: touch.clientY,
          width: size.width,
          height: size.height,
          posX: position.x,
          posY: position.y
        };
        document.body.style.userSelect = "none";
        e.stopPropagation();
      };
      return { onMouseDown, onTouchStart };
    },
    [size, position, bringToFront]
  );
  const resizeHandlers = useMemo(
    () => ({
      topLeft: createResizeHandlers("topLeft"),
      top: createResizeHandlers("top"),
      topRight: createResizeHandlers("topRight"),
      right: createResizeHandlers("right"),
      bottomRight: createResizeHandlers("bottomRight"),
      bottom: createResizeHandlers("bottom"),
      bottomLeft: createResizeHandlers("bottomLeft"),
      left: createResizeHandlers("left")
    }),
    [createResizeHandlers]
  );
  const handleClose = useCallback(() => {
    if (onClose) {
      return onClose();
    }
    setIsVisible(false);
  }, [onClose]);
  const clampWidth = useCallback(
    (w) => {
      let clamped = Math.max(minWidth, w);
      if (maxWidth !== void 0) {
        clamped = Math.min(maxWidth, clamped);
      }
      return clamped;
    },
    [minWidth, maxWidth]
  );
  const clampHeight = useCallback(
    (h) => {
      let clamped = Math.max(minHeight, h);
      if (maxHeight !== void 0) {
        clamped = Math.min(maxHeight, clamped);
      }
      return clamped;
    },
    [minHeight, maxHeight]
  );
  const applyDragBounds = useCallback(
    (newX, newY) => {
      let boundedX = newX;
      let boundedY = newY;
      if (dragBounds) {
        if (dragBounds.minX !== void 0) {
          boundedX = Math.max(dragBounds.minX, boundedX);
        }
        if (dragBounds.maxX !== void 0) {
          boundedX = Math.min(dragBounds.maxX, boundedX);
        }
        if (dragBounds.minY !== void 0) {
          boundedY = Math.max(dragBounds.minY, boundedY);
        }
        if (dragBounds.maxY !== void 0) {
          boundedY = Math.min(dragBounds.maxY, boundedY);
        }
      } else if (withinPortal) {
        boundedX = Math.max(0, Math.min(boundedX, window.innerWidth - size.width));
        boundedY = Math.max(0, Math.min(boundedY, window.innerHeight - 50));
      } else {
        const parent = windowRef.current?.offsetParent;
        if (parent instanceof HTMLElement) {
          const parentWidth = parent.clientWidth;
          const parentHeight = parent.clientHeight;
          boundedX = Math.max(0, Math.min(boundedX, parentWidth - size.width));
          boundedY = Math.max(0, Math.min(boundedY, parentHeight - 50));
        }
      }
      return { x: boundedX, y: boundedY };
    },
    [dragBounds, withinPortal, size.width]
  );
  const handleResize = useCallback(
    (clientX, clientY) => {
      const deltaX = clientX - resizeStart.current.x;
      const deltaY = clientY - resizeStart.current.y;
      let newWidth = size.width;
      let newHeight = size.height;
      let newX = resizeStart.current.posX;
      let newY = resizeStart.current.posY;
      switch (resizeDirection.current) {
        case "topLeft":
          newWidth = clampWidth(resizeStart.current.width - deltaX);
          newHeight = clampHeight(resizeStart.current.height - deltaY);
          newX = resizeStart.current.posX + (resizeStart.current.width - newWidth);
          newY = resizeStart.current.posY + (resizeStart.current.height - newHeight);
          break;
        case "top":
          newHeight = clampHeight(resizeStart.current.height - deltaY);
          newY = resizeStart.current.posY + (resizeStart.current.height - newHeight);
          break;
        case "topRight":
          newWidth = clampWidth(resizeStart.current.width + deltaX);
          newHeight = clampHeight(resizeStart.current.height - deltaY);
          newY = resizeStart.current.posY + (resizeStart.current.height - newHeight);
          break;
        case "right":
          newWidth = clampWidth(resizeStart.current.width + deltaX);
          break;
        case "bottomRight":
          newWidth = clampWidth(resizeStart.current.width + deltaX);
          newHeight = clampHeight(resizeStart.current.height + deltaY);
          break;
        case "bottom":
          newHeight = clampHeight(resizeStart.current.height + deltaY);
          break;
        case "bottomLeft":
          newWidth = clampWidth(resizeStart.current.width - deltaX);
          newHeight = clampHeight(resizeStart.current.height + deltaY);
          newX = resizeStart.current.posX + (resizeStart.current.width - newWidth);
          break;
        case "left":
          newWidth = clampWidth(resizeStart.current.width - deltaX);
          newX = resizeStart.current.posX + (resizeStart.current.width - newWidth);
          break;
      }
      setSize({ width: newWidth, height: newHeight });
      if (newX !== resizeStart.current.posX || newY !== resizeStart.current.posY) {
        setPosition({ x: newX, y: newY });
      }
    },
    [size.width, size.height, clampWidth, clampHeight, setSize, setPosition]
  );
  useEffect(() => {
    const handleMouseMove = (e) => {
      if (isDragging.current) {
        const newPos = applyDragBounds(
          e.clientX - dragStart.current.x,
          e.clientY - dragStart.current.y
        );
        setPosition(newPos);
      }
      if (isResizing.current) {
        handleResize(e.clientX, e.clientY);
      }
    };
    const handleTouchMove = (e) => {
      if (isDragging.current || isResizing.current) {
        const touch = e.touches[0];
        if (isDragging.current) {
          const newPos = applyDragBounds(
            touch.clientX - dragStart.current.x,
            touch.clientY - dragStart.current.y
          );
          setPosition(newPos);
        }
        if (isResizing.current) {
          handleResize(touch.clientX, touch.clientY);
        }
        e.preventDefault();
      }
    };
    const handleMouseUp = () => {
      if (isDragging.current || isResizing.current) {
        isDragging.current = false;
        isResizing.current = false;
        document.body.style.userSelect = "";
        document.body.style.cursor = "";
      }
    };
    const handleTouchEnd = () => {
      if (isDragging.current || isResizing.current) {
        isDragging.current = false;
        isResizing.current = false;
        document.body.style.userSelect = "";
        document.body.style.cursor = "";
      }
    };
    document.addEventListener("mousemove", handleMouseMove);
    document.addEventListener("mouseup", handleMouseUp);
    document.addEventListener("touchmove", handleTouchMove, { passive: false });
    document.addEventListener("touchend", handleTouchEnd);
    document.addEventListener("touchcancel", handleTouchEnd);
    return () => {
      document.removeEventListener("mousemove", handleMouseMove);
      document.removeEventListener("mouseup", handleMouseUp);
      document.removeEventListener("touchmove", handleTouchMove);
      document.removeEventListener("touchend", handleTouchEnd);
      document.removeEventListener("touchcancel", handleTouchEnd);
      document.body.style.userSelect = "";
      document.body.style.cursor = "";
    };
  }, [applyDragBounds, handleResize, setPosition]);
  return {
    isCollapsed,
    setIsCollapsed,
    isVisible,
    setIsVisible,
    zIndex,
    position,
    size,
    windowRef,
    handleMouseDownDrag,
    handleTouchStartDrag,
    resizeHandlers,
    handleClose,
    bringToFront
  };
}

export { useMantineWindow };
